File: ios/Sources/Design/Tokens.swift
import SwiftUI

/// Les mêmes jetons que la version web. Une seule identité visuelle, deux
/// plateformes. Rien n'est écrit en dur ailleurs dans l'app.
enum Token {
    enum Palette {
        static let background = Color(hex: 0xF5F5F7)
        static let card = Color.white
        static let elevated = Color(hex: 0xFBFBFD)
        static let text = Color(hex: 0x1D1D1F)
        static let secondary = Color(hex: 0x4B4B50)
        static let tertiary = Color(hex: 0x6E6E73)
        static let blue = Color(hex: 0x0066CC)
        static let green = Color(hex: 0x1F7A2E)
        static let orange = Color(hex: 0xB04503)
        static let red = Color(hex: 0xD10000)
        static let violet = Color(hex: 0x5E5CE6)
        static let hairline = Color.black.opacity(0.08)
        static let sunken = Color.black.opacity(0.025)
    }

    enum Space {
        static let xs: CGFloat = 4
        static let sm: CGFloat = 8
        static let md: CGFloat = 12
        static let lg: CGFloat = 16
        static let xl: CGFloat = 24
        static let xxl: CGFloat = 32
    }

    enum Radius {
        static let card: CGFloat = 18
        static let small: CGFloat = 12
        static let field: CGFloat = 10
    }

    /// La courbe d'apple.com : elle démarre doucement et s'arrête longuement.
    static let ease = Animation.timingCurve(0.28, 0.11, 0.32, 1, duration: 0.3)
}

extension Color {
    init(hex: UInt32) {
        self.init(
            .sRGB,
            red: Double((hex >> 16) & 0xFF) / 255,
            green: Double((hex >> 8) & 0xFF) / 255,
            blue: Double(hex & 0xFF) / 255,
            opacity: 1
        )
    }
}

extension View {
    /// La carte du produit : un filet, pas une ombre. L'ombre est réservée à
    /// ce qui est réellement au-dessus, une feuille ou une alerte.
    func cardSurface(padding: CGFloat = Token.Space.lg) -> some View {
        self
            .padding(padding)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(Token.Palette.card)
            .clipShape(RoundedRectangle(cornerRadius: Token.Radius.card, style: .continuous))
            .overlay(
                RoundedRectangle(cornerRadius: Token.Radius.card, style: .continuous)
                    .stroke(Token.Palette.hairline, lineWidth: 1)
            )
    }
}

File: android/app/src/main/java/app/capmedia/visaflow/design/Tokens.kt
package app.capmedia.visaflow.design

import androidx.compose.animation.core.CubicBezierEasing
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * Les mêmes jetons que l'iOS et que le web. Une seule identité visuelle, trois
 * plateformes. Rien n'est écrit en dur ailleurs dans l'application : une
 * couleur posée à la main dans un écran est une dette, pas un raccourci.
 *
 * Valeurs reprises telles quelles de ios/Sources/Design/Tokens.swift.
 */
object Token {

    object Palette {
        val background = Color(0xFFF5F5F7)
        val card = Color.White
        val elevated = Color(0xFFFBFBFD)
        val text = Color(0xFF1D1D1F)
        val secondary = Color(0xFF4B4B50)
        val tertiary = Color(0xFF6E6E73)
        val blue = Color(0xFF0066CC)
        val green = Color(0xFF1F7A2E)
        val orange = Color(0xFFB04503)
        val red = Color(0xFFD10000)
        val violet = Color(0xFF5E5CE6)
        val hairline = Color(0x14000000)
        val sunken = Color(0x06000000)
    }

    object Space {
        val xs: Dp = 4.dp
        val sm: Dp = 8.dp
        val md: Dp = 12.dp
        val lg: Dp = 16.dp
        val xl: Dp = 24.dp
        val xxl: Dp = 32.dp
    }

    object Radius {
        val card = 18.dp
        val small = 12.dp
        val field = 10.dp
    }

    /**
     * La courbe d'apple.com. Elle démarre doucement et s'arrête longuement,
     * ce qui est exactement l'inverse d'une courbe Material par défaut.
     */
    val easing = CubicBezierEasing(0.28f, 0.11f, 0.32f, 1f)

    fun <T> ease(durationMillis: Int = 300) = tween<T>(durationMillis, easing = easing)
}

/**
 * La carte du produit : un filet, jamais une ombre. L'ombre est réservée à ce
 * qui est réellement au-dessus, une feuille ou une alerte. C'est la règle qui
 * distingue le plus vite une interface Apple d'une interface Material.
 */
@Composable
fun CardSurface(
    modifier: Modifier = Modifier,
    padding: Dp = Token.Space.lg,
    spacing: Dp = Token.Space.md,
    content: @Composable androidx.compose.foundation.layout.ColumnScope.() -> Unit,
) {
    androidx.compose.foundation.layout.Column(
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(Token.Radius.card))
            .background(Token.Palette.card)
            .border(1.dp, Token.Palette.hairline, RoundedCornerShape(Token.Radius.card))
            .padding(padding),
        verticalArrangement = Arrangement.spacedBy(spacing),
        content = content,
    )
}

/** Les gouttières de l'écran, identiques partout. */
val ScreenPadding = PaddingValues(horizontal = 16.dp, vertical = 16.dp)

object Type {
    val displaySize = 34.sp
    val titleSize = 22.sp
    val headlineSize = 17.sp
    val bodySize = 16.sp
    val captionSize = 13.sp
    val microSize = 11.sp
}
